### ISSUE TITLE
If searcher zip archives are corrupted, we should gracefully handle it

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#1759]
It is possible to e.g. shutdown a searcher instance while a zip archive from gitserver is being transferred to it, resulting in empty or corrupt zip files on disk. This leads to search errors like:

```
1 error occurred: * text search failed: failed to search stash.example.com/myorg/myrepo@: zip: not a valid zip file
```

The workaround is to find and remove the corrupt zip by running on each searcher instance a command that will identify the corrupt zips and delete them:

```
find /mnt/cache/searcher/searcher-archives -type f | xargs -I % sh -c 'unzip -lqq % >/dev/null 2>&1 || (echo "removing corrupt archive %" && rm %)'
```

We should instead automatically do this so our users do not have to. i.e., if a zip archive is invalid we should request it from gitserver again.

cc @keegancsmith

**Comment by keegancsmith** [comment]
This shouldn't happen, we download zip files to a tmp location first, then once the download is complete we do an atomic rename operation. However, we are possibly missing out on some syncs/flushes. Some filesystems can sync metadata changes (like the file renames) before data, which on restart can lead to a weird state.

As such we should throw away zip files that fail to open and return the same error we would return for zip files that don't download in time.

https://github.com/sourcegraph/sourcegraph/blob/8065e2c2bd34866f8561e467f7aac27d6a59e9c4/pkg/diskcache/cache.go#L150-L180

**Comment by nicksnyder** [comment]
This is less of a priority than https://github.com/sourcegraph/sourcegraph/issues/2025 so I am backlogging.

**Comment by slimsag** [comment]
This has happened to actual customers and required manual intervention, which is bad. I agree the priority is less than #2025 but we need to address this soon. Let's aim for 3.2?

**Comment by sqs** [comment]
Removing milestone to reflect that we need to prioritize this.

**Comment by sqs** [comment]
Reassigning to @ijsnow 	because this is search related, and @ijsnow 	is working on search backend. (If you disagree, feel free to reassign it back.)

@ijsnow Because #2081 says "The rest of the search issues in the milestone can be prioritized after the above issues are done.", I am adding this to the 3.2 milestone and labeling it with `search`. This issue is more important than those described by "rest of the search issues".

**Comment by nicksnyder** [comment]
Let's avoid phrases like `The rest of the search issues in the milestone can be prioritized after the above issues are done.` in tracking issues in the future. It encourages packing more things into the milestone than are possible. If something is expected to get done then it should be explicitly enumerated in the tracking issue, otherwise it should be backlog.

**Comment by ijsnow** [comment]
@slimsag do you have steps to reproduce this?

**Comment by slimsag** [comment]
@ijsnow no, if you want to reproduce it then just corrupt one of the archive files by writing random bytes into the file so it's not a valid zip.

**Comment by slimsag** [comment]
Moving to 3.4 since this didn't get done

**Comment by slimsag** [comment]
@sourcegraph/code-search I've just encountered this again, can you re-prioritize? This time on my https://app.hubspot.com/contacts/2762526/company/407948923 replica instance, but with the tar archives that searcher uses instead of the zip archives. The corruption occurred during a redeployment of my instance.

```
1 error occurred: * text search failed: failed to search 50.112.230.151/TEST/ssw-scientific--nndd@: failed to fetch missing archive cache item: failed to copy and close missing archive cache item: archive/tar: invalid tar header
```

This error comes from here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@95fb2bcf39e72dc9ae8e5328abcf8bcffbabd909/-/blob/cmd/searcher/search/store.go#L217

It also appears that despite this error only affecting that one repository, we outright fail the entire search query when this happens. We should fix this too.

i.e. this doesn't produce any results due to the error above:

```
type:file err
```

But this does:

```
type:file err -repo:TEST\/ssw-scientific
```

**Comment by slimsag** [comment]
OK, I found out the cause here is not exactly a corrupt cache, it just looks like one. I will file issues separately.

- Searcher fetches tars from Gitserver and then translates them into the zip files for storage on disk.
- `FetchTar` is defined for searcher here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@95fb2bc/-/blob/cmd/searcher/main.go#L53-55
- Calls to `FetchTar` are effectively wrapped in a `*pkg/diskcache.Store`: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@95fb2bc/-/blob/cmd/searcher/search/store.go#L145-147
- The error comes from calling the fetcher: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@95fb2bc/-/blob/pkg/diskcache/cache.go#L187 So the error comes directly from calling https://sourcegraph.com/github.com/sourcegraph/sourcegraph@95fb2bc/-/blob/cmd/searcher/search/store.go#L146 not from any cached state as part of the diskcache. **The bug isn't in diskcache.**

gitserver is failing to clone the repo:

```
docker logs gitserver-16 2>&1 | grep 'ssw-scientific--nndd'
t=2019-05-02T20:20:21+0000 lvl=info msg="recloning expired repo" repo=50.112.230.151/TEST/ssw-scientific--nndd
t=2019-05-02T20:20:24+0000 lvl=eror msg="error running cleanup command" name="maybe reclone" repo=/data/repos/50.112.230.151/TEST/ssw-scientific--nndd/.git error="error cloning repo: repo 50.112.230.151/TEST/ssw-scientific--nndd (http://slimsag:<redacted>@10.0.0.34:7990/scm/test/ssw-scientific--nndd.git) not cloneable: exit status 128 (output follows)\n\nfatal: unable to access 'http://slimsag:<redacted>@10.0.0.34:7990/scm/test/ssw-scientific--nndd.git/': Failed to connect to 10.0.0.34 port 7990: Host is unreachable\n"
```

So, the problems here (revised) are:

1. https://github.com/sourcegraph/sourcegraph/issues/3798
2. https://github.com/sourcegraph/sourcegraph/issues/3799

Closing in favor of those.
